deps/rabbitmq_mqtt/src/rabbit_mqtt_retained_msg_store_dets.erl

```diff
@@ -10,60 +10,288 @@
+    Expire = rabbit_mqtt_retained_msg_store:expire(dets, MsgTable),
+    {ok, NodeTable} = open_table(Dir, VHost, <<"nodes">>, set),
+    {ok, EdgeTable} = open_table(Dir, VHost, <<"edges">>, set),
+    {ok, RootId} = find_or_insert_root_node(NodeTable, EdgeTable),
+    MaxRetainedMessagesCount =
+      rabbit_mqtt_retained_msg_store:get_max_retained_messages_count(),
+    State =
+      #store_state{node_table = NodeTable,
+                   edge_table = EdgeTable,
+                   msg_table = MsgTable,
+                   root_id = RootId,
+                   dir = Dir,
+                   vhost = VHost,
+                   max_retained_messages_count = MaxRetainedMessagesCount},
+    {ok, State, Expire}
+  catch
+    error:Reason ->
+      ?LOG_ERROR("Failed to recover MQTT retained message store: ~p", [Reason]),
+      {error, uninitialized}
+  end.
 
 -spec insert(topic(), mqtt_msg(), store_state()) -> ok.
-insert(Topic, Msg, #store_state{table = T}) ->
-  ok = dets:insert(T, #retained_message{topic = Topic, mqtt_msg = Msg}).
-
--spec lookup(topic(), store_state()) ->
-    mqtt_msg() | mqtt_msg_v0() | undefined.
-lookup(Topic, #store_state{table = T}) ->
-  case dets:lookup(T, Topic) of
-    []      -> undefined;
-    [#retained_message{mqtt_msg = Msg}] -> Msg
-  end.
+insert(Topic, Msg, #store_state{} = State) ->
+  Words = split_topic(Topic),
+  NodeId = follow_or_create_path(Words, State),
+  % Mark node as topic end and store message
+  update_node(NodeId, true, State),
+  dets:insert(State#store_state.msg_table, {NodeId, Topic, Msg}),
+  ok.
+
+-spec lookup(topic(), store_state()) -> [mqtt_msg()] | [mqtt_msg_v0()] | [].
+lookup(Topic,
+       #store_state{root_id = RootId,
+                    msg_table = MsgTable,
+                    max_retained_messages_count = Limit} =
+         State) ->
+  Words = split_topic(Topic),
+  Matches = lists:sublist(match_pattern_words(Words, RootId, State, []), Limit),
+  Values =
+    lists:flatmap(fun(NodeId) ->
+                     case dets:lookup(MsgTable, NodeId) of
+                       [] -> [];
+                       [{_NodeId, _Topic, Value} | _] -> [Value];
+                       {error, _Reason} ->
+                         ?LOG_ERROR("Failed to lookup MQTT retained message for node ~p", [NodeId]),
+                         []
+                     end
+                  end,
+                  Matches),
+  Values.
 
 -spec delete(topic(), store_state()) -> ok.
-delete(Topic, #store_state{table = T}) ->
-  ok = dets:delete(T, Topic).
+delete(Topic, State) ->
+  Words = split_topic(Topic),
+  case follow_path(Words, State) of
+    {ok, NodeId} ->
+      dets:match_delete(State#store_state.msg_table, {NodeId, Topic, '_'}),
+      case dets:lookup(State#store_state.msg_table, NodeId) of
+        [] ->
+          update_node(NodeId, false, State),
+          maybe_clean_path(NodeId, State);
+        _ ->
+          ok
+      end;
+    error ->
+      ok
+  end,
+  ok.
```

Comment:
it looks like I got the deletion logic wrong: first i should update the node, cleanup edges and then delete the message from the msg_table.

Reply:
also delete_match is probably pointless after moving from bag to set.